gui/src/app/pages/HomePage/HomePage.tsx

```diff
@@ -30,23 +30,39 @@ const HomePage: FunctionComponent<Props> = ({ width, height }) => {
 }
 
 const HomePageChild: FunctionComponent<Props> = ({ width, height }) => {
-
-
     const { data, update } = useContext(SPAnalysisContext)
     const setSamplingOpts = useCallback((opts: SamplingOpts) => {
         update({ type: 'setSamplingOpts', opts })
     }, [update])
 
     const [compiledMainJsUrl, setCompiledMainJsUrl] = useState<string>('')
 
-    const leftPanelWidth = Math.max(250, Math.min(340, width * 0.2))
+    const [leftPanelCollapsed, setLeftPanelCollapsed] = useState(determineShouldBeInitiallyCollapsed(width))
+    const expandedLeftPanelWidth = determineLeftPanelWidth(width) // what the width would be if expanded
+    const leftPanelWidth = leftPanelCollapsed ? 20 : expandedLeftPanelWidth // the actual width
+
+    // We automatically collapse the panel if user has resized the window to be
+    // too small but we only want to do this right when we cross the threshold,
+    // not every time we resize by a pixel. Similar for expanding the panel when
+    // we cross the threshold in the other direction.
+    const lastWidth = useRef(width)
+    useEffect(() => {
+        if (!determineShouldBeInitiallyCollapsed(lastWidth.current) && determineShouldBeInitiallyCollapsed(width)) {
+            lastWidth.current = width
+            setLeftPanelCollapsed(true)
+        }
+        else if (determineShouldBeInitiallyCollapsed(lastWidth.current) && !determineShouldBeInitiallyCollapsed(width)) {
+            lastWidth.current = width
+            setLeftPanelCollapsed(false)
+        }
+    }, [width])
```

Comment:
Since all we care about for `lastWidth` is whether it implies a collapsed state, this could probably be simplified (reducing function calls) as something along the lines of:
```Typescript
const isCollapsed = useRef(determineShouldBeInitiallyCollapsed(width));
useEffect(() => {
  const shouldBeCollapsed = determineShouldBeInitiallyCollapsed(width);
  if (isCollapsed.current !== shouldBeCollapsed) {
    isCollapsed.current = shouldBeCollapsed;
    setLeftPanelCollapsed(shouldBeCollapsed);
  } 
}, [width])
```

Reply:
I think your `isCollapsed` variable should really be named something like `lastShouldBeCollapsed`. Either way it's somewhat confusing. I think I'd prefer to stick with what I have.

Reply:
Thinking more about it, I think your simplification is clearer. I essentially used your code but I used the name `lastShouldBeCollapsed`